File: contingency.py
import sys
import numpy as np
from sets import Set
from ale_python_interface import ALEInterface


def getRAMVector(ale):
  ram_size = ale.getRAMSize()
  ram = np.zeros((ram_size,), dtype=np.uint8)
  ale.getRAM(ram)
  return ram


def initializeALE(romFile):
  ale = ALEInterface()

  ale.setInt("max_num_frames_per_episode", 18000)
  ale.setInt("random_seed", 123)
  ale.setFloat("repeat_action_probability", 0.0)
  ale.setInt("frame_skip", 5)

  random_seed = ale.getInt("random_seed")
  print("random_seed: " + str(random_seed))

  # Set USE_SDL to true to display the screen. ALE must be compilied
  # with SDL enabled for this to work. On OSX, pygame init is used to
  # proxy-call SDL_main.

  USE_SDL = False
  if USE_SDL:
    if sys.platform == 'darwin':
      import pygame
      pygame.init()
      ale.setBool('sound', False) # Sound doesn't work on OSX
    elif sys.platform.startswith('linux'):
      ale.setBool('sound', True)
    ale.setBool('display_screen', True)

  ale.loadROM(romFile)
  actionSet = ale.getMinimalActionSet()

  return ale, actionSet


if __name__ == "__main__":
  # Read arguments
  if(len(sys.argv) < 2):
    print("Usage ./ale_python_test1.py <ROM_FILE_NAME>")
    sys.exit()

  # Initialize ALE
  ale, actionSet = initializeALE(sys.argv[1])
  prevRAM = getRAMVector(ale)

  changing_bytes = Set([])
  len_changing_bytes = []
  contingent_bytes = Set([])
  len_contingent_bytes = []
  while not ale.game_over():
    # Try all actions 
    ram_diffs = []
    for actionIdx in xrange(actionSet.size):
      ale.saveState()
      reward = ale.act(actionSet[actionIdx]);
      currRAM = getRAMVector(ale)
      ram_diffs.append(prevRAM - currRAM)
      ale.loadState()

    # Test to see which bytes vary according to the action taken
    for i in xrange(ale.getRAMSize()): # Iterate over each byte
      contingent = False
      # Pairwise combination of different actions
      for j in xrange(len(ram_diffs)):
        for k in xrange(len(ram_diffs)):
          if j < k and int(ram_diffs[j][i]) - int(ram_diffs[k][i]) != 0:
            contingent = True
      if contingent:
        contingent_bytes.add(i)

    # Test to see which bytes vary, regardless of contingency
    for i in xrange(ale.getRAMSize()): # Iterate over each byte
      for j in xrange(len(ram_diffs)):
        if ram_diffs[j][i] != 0:
          changing_bytes.add(i)

    # Random walk
    a = actionSet[np.random.randint(actionSet.size)]
    reward = ale.act(a);
    prevRAM = currRAM

    # Keeping track of the evolving length
    len_contingent_bytes.append(len(contingent_bytes))
    len_changing_bytes.append(len(changing_bytes))


  import matplotlib.pylab as plt
  plt.title("Space Invaders")
  plt.ylabel('Set size')
  plt.xlabel('Time steps (w/ frame skip 5)')
  plt.plot(len_contingent_bytes, label='Contingent bytes')
  plt.plot(len_changing_bytes, label='Changing bytes')
  plt.legend()
  plt.savefig('graphs/space_invaders.png')

File: contingency.cpp
/* *****************************************************************************
 * Author: Marlos C. Machado
 *
 *  Sample code for computing the number of contingent bytes in a game.
 **************************************************************************** */

#include <set>
#include <iostream>
#include <ale_interface.hpp>

#ifdef __USE_SDL
  #include <SDL.h>
#endif

using namespace std;

std::vector<int> subtractRAMs(ALERAM ram1, ALERAM ram2) {
    assert(ram1.size() == ram2.size());
    std::vector<int> result(ram1.size());
    for(unsigned i = 0; i < ram1.size(); i++) {
        result[i] = ram1.get(i) - ram2.get(i);
    }
    return result;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " rom_file" << std::endl;
        return 1;
    }

    ALEInterface ale;

    // Get & Set the desired settings
    ale.setInt("frame_skip", 5);
    ale.setInt("random_seed", 123);
    ale.setInt("max_num_frames_per_episode", 18000);
    ale.setFloat("repeat_action_probability", 0.0); // we are doing planning

#ifdef __USE_SDL
    ale.setBool("display_screen", true);
    ale.setBool("sound", true);
#endif

    // Load the ROM file. (Also resets the system for new settings to take effect.)
    ale.loadROM(argv[1]);

    // Get the vector of legal actions
    ActionVect actions = ale.getLegalActionSet();

    int reward;
    ALERAM prev_ram, curr_ram;
    prev_ram = ale.getRAM();

    std::set<int> changing_bytes;
    std::vector<int> len_changing_bytes;
    std::set<int> contingent_bytes;
    std::vector<int> len_contingent_bytes;
    std::vector<std::vector<int>> difference_rams;

    while(!ale.game_over()) {
        // Try all actions
        for(unsigned a = 0; a < actions.size(); a++) {
            ale.saveState();
            reward = ale.act(actions[a]);
            curr_ram = ale.getRAM();
            difference_rams.push_back(subtractRAMs(prev_ram, curr_ram));
            ale.loadState();
        }

        // Test to see which bytes vary according to the action taken
        for(unsigned i = 0; i < curr_ram.size(); i++) {
            bool contingent = false;
            // Pairwise combination of different actions
            for(unsigned j = 0; j < difference_rams.size(); j++) {
                for(unsigned k = 0; k < difference_rams.size(); k++) {
                    if(j < k && difference_rams[j][i] - difference_rams[k][i] != 0) {
                        contingent = true;
                    }
                }
            }
            if(contingent == true){
                contingent_bytes.insert(i);
            }
        }

        // Test to see which bytes vary, regardless of contingency
        for(unsigned i = 0; i < curr_ram.size(); i++) {
            for(unsigned j = 0; j < difference_rams.size(); j++) {
                if(difference_rams[j][i] != 0) {
                    changing_bytes.insert(i);
                }
            }
        }
        // Random walk
        Action a = actions[rand() % actions.size()];
        reward = ale.act(a);
        prev_ram = curr_ram;

        // Keeping track of the evolving length
        len_contingent_bytes.push_back(contingent_bytes.size());
        len_changing_bytes.push_back(changing_bytes.size());

        difference_rams.clear();
    }

    std::cout << "contingent_bytes = [";
    for(unsigned i = 0; i < len_contingent_bytes.size(); i++) {
        std::cout << len_contingent_bytes[i] << ", ";
    }
    std::cout << "]\nchanging_bytes = [";
    for(unsigned i = 0; i < len_changing_bytes.size(); i++) {
        std::cout << len_changing_bytes[i] << ", ";
    }
    std::cout << "]\n";

    return 0;
}
